db/recovery_db.go

```diff
@@ -0,0 +1,108 @@
+package db
+
+import (
+	"encoding/json"
+
+	"github.com/athanorlabs/atomic-swap/common"
+	"github.com/athanorlabs/atomic-swap/common/types"
+	mcrypto "github.com/athanorlabs/atomic-swap/crypto/monero"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+
+	"github.com/ChainSafe/chaindb"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	recoveryPrefix             = "recv"
+	contractAddrPrefix         = "addr"
+	contractSwapInfoPrefix     = "contract"
+	swapPrivateKeyPrefix       = "privkey"
+	sharedSwapPrivateKeyPrefix = "sprivkey"
+)
+
+// RecoveryInfo ...
+type RecoveryInfo struct {
+	ContractAddress      string
+	ContractSwapID       [32]byte
+	ContractSwap         contracts.SwapFactorySwap
+	PrivateKeyInfo       *mcrypto.PrivateKeyInfo
+	SharedSwapPrivateKey *mcrypto.PrivateKeyInfo
+}
+
+// RecoveryDB contains information about ongoing swaps requires for recovery
+// in case of shutdown.
+type RecoveryDB struct {
+	db chaindb.Database
+}
+
+func newRecoveryDB(db chaindb.Database) *RecoveryDB {
+	return &RecoveryDB{
+		db: db,
+	}
+}
+
+func getRecoveryDBKey(id types.Hash, additional string) []byte {
+	return append(id[:], []byte(additional)...)
+}
+
+func (db *RecoveryDB) close() error {
+	return db.db.Close()
+}
+
+// PutContractAddress stores the given contract address for the given swap ID.
+func (db *RecoveryDB) PutContractAddress(id types.Hash, addr ethcommon.Address) error {
+	val, err := json.Marshal(addr)
+	if err != nil {
+		return err
+	}
+
+	key := getRecoveryDBKey(id, contractAddrPrefix)
+	return db.db.Put(key[:], val)
+}
+
+// PutContractSwapInfo stores the given contract swap ID (which is not the same as the daemon
+// swap ID, but is instead a hash of the `SwapFactorySwap` structure)
+// and contract swap structure for the given swap ID.
+func (db *RecoveryDB) PutContractSwapInfo(id types.Hash, swapID [32]byte, swap contracts.SwapFactorySwap) error {
+	val, err := json.Marshal(&struct {
+		SwapID [32]byte
+		Swap   contracts.SwapFactorySwap
```

Comment:
Can we pull this out, give it the semantic version field, add JSON annotations, use a hash type that will auto-serialize as hex, and add at least one unit test for the JSON marshaling so you can see what the JSON looks like when reading the code? I is a little painful, but the odds of the format never changing are probably low. Having the ID in hex will help us troubleshoot. I can update the PR later today if you want.

Reply:
updated, but I'm not sure if I see the point of adding the version necessarily since this info is likely not going to be persisted long-term - it's only stored while a swap is ongoing, and will end up being deleted after

Reply:
Yeah, then it isn't needed. I thought a swap might be created with one version of `swapd` and might possibly be recovered by a newer `swapd` version.  I didn't add it in `dimalinux/db-swap-info-json`.